🚨  database: change hard limits and field types

kirrg001 opened the pull request:
refs #7432

- we went over all schema fields and decided to decrease/increase the hard limits
- the core goal is to have more flexibility in the future
- we reconsidered string vs. text

There are 5 groups:

- small strings: 50 characters
    - static strings
    - status, visibility, language, role name, permission name, client name etc.
- medium strings: 191 characters
    - all unique fields or fields which can be unique in the future
    - slug, tokens, user name, password, tag name, email
- large strings: 1000-2000 characters
    - these fields need to be very flexible
    - these fields get a soft limit attached (in a different PR)
    - post title, meta title, meta description, urls
- text: 64kb characters
    - bio, settings, location, tour
- long text: 1000000000 chars
    - html, amp, mobiledoc, markdown

kirrg001 committed:
🚨  database: change hard limits and field types

refs #7432

- we went over all schema fields and decided to decrease/increase the hard limits
- the core goal is to have more flexibility in the future
- we reconsidered string vs. text

There are 5 groups:

- small strings: 50 characters
    - static strings
    - status, visibility, language, role name, permission name, client name etc.
- medium strings: 191 characters
    - all unique fields or fields which can be unique in the future
    - slug, tokens, user name, password, tag name, email
- large strings: 1000-2000 characters
    - these fields need to be very flexible
    - these fields get a soft limit attached (in a different PR)
    - post title, meta title, meta description, urls
- medium text: 64kb characters
    - bio, settings, location, tour
- long text: 1000000000 chars
    - html, amp, mobiledoc, markdown

kirrg001 committed:
🙄  sort_order for tests

- sort order was not set for the tests, so it was always 0
- mysql could return a different result

in my case:
- field length 156 returned the following related tags ["bacon", "kitchen"]
- field length 157 returned the following related tags ["kitchen", "kitchen"]

kirrg001 commented:
@ErisDS 
FYI: 

it's worth looking at the second commit
https://github.com/TryGhost/Ghost/pull/7932/commits/bcfdfcd9538b2ff4b810c63e3033e480cb796bd7

ErisDS commented:
I have some small questions / amends here, but I think this is totally the right lines:

- what is the significance of 191 again? It would be useful to have a reference / quick explanation on this PR.
- should `client.secret` be 191?
- shouldn't the `id` fields have a limit?  They are a fixed size, and otherwise it will default to 255 which is quite big & will run into problems if fields need to be 191!
- should `user.name` and `tag.name` be treated the same as `post.title`? (they are similar)
- I think `tag.description` and `client.description` should probably be the same class as `user.bio`, maybe `role.description` too for consistency although I can't really see a use case.
- I think it's totally fine to keep a couple of unique lengths as a "unique" class. E.g. language is always a string like `en_US` or `de_DE` so don't think it needs to be put in the 50 char class. Same for `uuid` which is always 36 chars and I think password digests are fixed length tooo?

kirrg001 commented:
> what is the significance of 191 again? It would be useful to have a reference / quick explanation on this PR.

Sure. 
In `1.0.0` we are using `utf8mb4` encoding. Using this encoding requires a max length of 191 characters for unique keys or indexes. https://dev.mysql.com/doc/refman/5.7/en/charset-unicode-conversion.html

So the group for 191 characters included
- fields, which are candidates for unique keys or indexes
- fields, which are unique keys or indexes
- fields which fit into this group, because they need a medium length of characters

> should client.secret be 191?

yes, sorry!

> shouldn't the id fields have a limit? They are a fixed size, and otherwise it will default to 255 which is quite big & will run into problems if fields need to be 191!

See https://github.com/TryGhost/Ghost/blob/master/core/server/data/schema/commands.js#L19
I agree, this is hard to guess, but when i changed the id's in #7494, i was lazy and added a fallback.

> should user.name and tag.name be treated the same as post.title? (they are similar)

My thinking was that a user name or a tag name are less flexible than a post title.
I would like to hear your opinion.

> I think tag.description and client.description should probably be the same class as user.bio, maybe role.description too for consistency although I can't really see a use case.

I'd suggest increasing client description to 1000. But making this as text with 65535 KB feels weird?

> I think it's totally fine to keep a couple of unique lengths as a "unique" class. E.g. language is always a string like en_US or de_DE so don't think it needs to be put in the 50 char class. Same for uuid which is always 36 chars and I think password digests are fixed length tooo?

Yeah thought about it. But at the end I decided to group all small strings into one static length. Just for consistency and flexibility, in case we need more space.

I am very open to **every** change :)

ErisDS committed:
Change client.secret to 191

ErisDS commented:
>> shouldn't the id fields have a limit? They are a fixed size, and otherwise it will default to 255 which is quite big & will run into problems if fields need to be 191!

> See https://github.com/TryGhost/Ghost/blob/master/core/server/data/schema/commands.js#L19
I agree, this is hard to guess, but when i changed the id's in #7494, i was lazy and added a fallback.

Ah OK, that's cool! 

Still, would it not be best if we set them to 24 chars? They're not going to be any bigger than that, ever.

>> I think it's totally fine to keep a couple of unique lengths as a "unique" class.
> Yeah thought about it. But at the end I decided to group all small strings into one static length. Just for consistency and flexibility, in case we need more space.

I see this both ways. For one consistency is nice where it's possible. It's certainly nice to have classes to pick from. However, if we care enough to shrink 150 chars to 50 - surely it's sensible to also restrict fields where they are 100% fixed? E.g. UUID, it's never ever ever gonna be more than 36 chars. 

Let me know if you have some knowledge that it is not important 👍  Otherwise I think it is good to make fixed length fields fixed, and use the "classes" everywhere else 😁

ErisDS committed:
Tweak field lengths

- Add 24 char limit for ids
- Limited fields are the exact length they need
- Unified 1000 and 2000 char string classes to all be 2000
- Changed descriptions to be either 2000, except user & tag which is text 65535 as these may be used to store HTML later?!
- Updated tests

ErisDS commented:
I have made some tweaks. I'm not 100% sure about the description fields / large string class yet - e.g. bio.  Still thinking on this for now, probably need to read a bit more about the differences between varchar vs text.

The use case is to allow people to start storing HTML in their bios & descriptions for things like authors and tags. Still I don't really see them getting very, very long. 

1 other question I have left - how long is the token in `user. ghost_auth_access_token`? I think this can have a fixed length.

kirrg001 commented:
Sorry for very late response!

> 1 other question I have left - how long is the token in user. ghost_auth_access_token? I think this can have a fixed length.

It's 32.

ErisDS committed:
🛠  Update importer tests

- The old 001-003 tests are kind of less relevant now.
- Rather than worrying about past versions of the data structure, we should check that the importer only imports what we consider to be valid data
- I've changed the tests to treat the title-length check as a length-validation check, rather than a test for each of the old versions

ErisDS commented:
OK. I've updated this to set the `ghost_auth_access_token` to 32, to pass the tests and to make importer tests pass.

I've got two concerns that have been playing on my mind: The first is that the medium length description type fields don't have clear enough use-cases that I am certain they are correct. The second is that we might run into horrible extreme-edge-case errors when importing with the fields that have gotten shorter and are the reasons for shortening fields worth the potential pain?

My conclusion is that we don't need to do anything with this PR, we've got the best setup we can have given the info we have. However, I there are two things left to do afterwards:

1. soft limits
2. make sure the importer errors are as clear as possible.

The changes I just made to the importer tests made one other thing apparent: at the moment we have quite weird tests where we tried to make sure that data from each database version would always import. This results in test duplication, and we promptly forgot to do it for 004, 005 etc - it doesn't really make sense.

Instead, we should probably define the constraints of what data we should import. E.g. a title can be up to 2000 chars long and contain utf8/utf8mb4 data. Then we should test the bounds of these constraints. As long as the data constraints stay the same, we won't need to write extra tests. If a constraint changes, we should write tests to ensure the import works correctly before and after the constraint change, however we deem "correctly" to be (e.g. truncate a value, or produce an error, etc).

Probably need to start collecting together ideas for importer changes into an issue or somewhere?

kirrg001 commented:
> Probably need to start collecting together ideas for importer changes into an issue or somewhere?

I have already created an importer note for the next sprint and added some details. I have also added your comment ^. Definitely worth an issue.

> we might run into horrible extreme-edge-case errors when importing with the fields 

As far as i can see the only fields which are shorter are static strings defined in Ghost e.g. status, role names etc. Tell me if i have overseen something.

> The first is that the medium length description type fields don't have clear enough use-cases that I am certain they are correct

The only use case in my head is, most of them are candidates for unique constraints.

> My conclusion is that we don't need to do anything with this PR

agree 👍

ErisDS commented:
> As far as i can see the only fields which are shorter are static strings defined in Ghost e.g. status, role names etc. Tell me if i have overseen something.

Nope, you are correct. Totally crazy edge case, I don't think we need to worry about it at all it was just in my head and I wanted to get it out somewhere 😝 

I think this PR is GTG?! WHOOOOP

kirrg001 commented on core/server/data/schema/schema.js:
```diff
@@ -1,21 +1,21 @@
     posts: {
-        id: {type: 'string', nullable: false, primary: true},
+        id: {type: 'string', maxlength: 24, nullable: false, primary: true},
         uuid: {type: 'string', maxlength: 36, nullable: false, validations: {isUUID: true}},
```
uuid is here 36 and for client it's 50. on purpose?
i don't remember.

ErisDS replied:
No, missed that, fixing

kirrg001 commented on core/server/data/schema/schema.js:
```diff
@@ -52,90 +52,90 @@ module.exports = {
-        id: {type: 'string', nullable: false, primary: true},
+        id: {type: 'string', maxlength: 24, nullable: false, primary: true},
         role_id: {type: 'string', nullable: false},
         user_id: {type: 'string', nullable: false}
```
I can see some field which don't have the 24 length definition.
e.g. created_by, user_id

ErisDS replied:
Fuck, fixing.

ErisDS commented:
Is the schema for brute definitely correct? I did not review this. cc @cobbspur as well here.

cobbspur commented:
I believe it took @kirrg001 a while to nail down the brute schema, don't remember all the ins and outs but it is what are using elsewhere and is working

ErisDS committed:
Fix more maxlengths

ErisDS committed:
🔥 Remove foreign key from subscribers.post_id

- There's no real need to have an index on this column, it just makes deleting posts hard.
- Same as created_by type columns, we can reference ids without needing keys/indexes
